File: campus-notifications/stage1/priority_inbox.py
"""
Stage 1: Priority Inbox - Campus Notifications Microservice
Fetches notifications from API and returns top N by priority score.
Priority = weight(type) + recency_score(timestamp)
"""

import requests
import heapq
from datetime import datetime
from typing import Optional
import argparse
import json
import os

# ─── Configuration ────────────────────────────────────────────────────────────
API_URL = "http://4.224.186.213/evaluation-service/notifications"

# Weight by notification type: Placement > Event > Result
TYPE_WEIGHT = {
    "Placement": 3,
    "Event": 2,
    "Result": 1,
}

# ─── Auth / Headers ───────────────────────────────────────────────────────────
def get_headers() -> dict:
    """
    The API is a protected route.
    Add your Bearer token to env var NOTIFICATION_API_TOKEN if required.
    """
    token = os.environ.get("NOTIFICATION_API_TOKEN", "")
    headers = {"Content-Type": "application/json"}
    if token:
        headers["Authorization"] = f"Bearer {token}"
    return headers


# ─── Fetch ────────────────────────────────────────────────────────────────────
def fetch_notifications() -> list[dict]:
    """Fetch all notifications from the evaluation-service API."""
    try:
        resp = requests.get(API_URL, headers=get_headers(), timeout=10)
        resp.raise_for_status()
        data = resp.json()
        return data.get("notifications", [])
    except requests.exceptions.RequestException as e:
        print(f"[ERROR] Failed to fetch notifications: {e}")
        return []


# ─── Scoring ─────────────────────────────────────────────────────────────────
def parse_timestamp(ts: str) -> datetime:
    """Parse ISO-like timestamp string."""
    try:
        return datetime.strptime(ts, "%Y-%m-%d %H:%M:%S")
    except ValueError:
        return datetime.min


def compute_priority_score(notification: dict, now: datetime) -> float:
    """
    Priority Score = type_weight + recency_score

    recency_score: higher for more recent notifications.
      We normalise age in minutes; cap at 1440 min (24 h).
      recency_score = max(0, 1 - age_minutes / 1440)
      This gives a value in [0, 1].

    Final score is therefore in range (0, 4].
    """
    type_w = TYPE_WEIGHT.get(notification.get("Type", ""), 1)
    ts = parse_timestamp(notification.get("Timestamp", ""))
    age_minutes = max(0, (now - ts).total_seconds() / 60)
    recency = max(0.0, 1.0 - age_minutes / 1440.0)
    return type_w + recency


# ─── Core Logic ───────────────────────────────────────────────────────────────
def get_top_n_notifications(n: int = 10, notifications: Optional[list] = None) -> list[dict]:
    """
    Return the top-n priority notifications from the live API (or a supplied list).
    Uses a min-heap of size n for O(M log n) efficiency where M = total notifications.
    """
    if notifications is None:
        notifications = fetch_notifications()

    if not notifications:
        return []

    now = datetime.now()

    # heap stores (score, index, notification) — index breaks ties deterministically
    heap: list[tuple] = []

    for idx, notif in enumerate(notifications):
        score = compute_priority_score(notif, now)
        if len(heap) < n:
            heapq.heappush(heap, (score, idx, notif))
        elif score > heap[0][0]:
            heapq.heapreplace(heap, (score, idx, notif))

    # Sort descending by score
    top = sorted(heap, key=lambda x: x[0], reverse=True)
    return [item[2] for item in top]


# ─── Display ──────────────────────────────────────────────────────────────────
def display_priority_inbox(top_notifications: list[dict], n: int) -> None:
    """Pretty-print the priority inbox."""
    print(f"\n{'═' * 60}")
    print(f"  🔔  PRIORITY INBOX  —  Top {n} Notifications")
    print(f"{'═' * 60}")

    if not top_notifications:
        print("  No notifications found.")
        print(f"{'═' * 60}\n")
        return

    type_icons = {"Placement": "💼", "Event": "🎉", "Result": "📊"}

    for rank, notif in enumerate(top_notifications, start=1):
        icon = type_icons.get(notif.get("Type", ""), "🔔")
        print(f"\n  #{rank}  {icon}  [{notif.get('Type', 'Unknown')}]")
        print(f"       Message  : {notif.get('Message', 'N/A')}")
        print(f"       Timestamp: {notif.get('Timestamp', 'N/A')}")
        print(f"       ID       : {notif.get('ID', 'N/A')}")

    print(f"\n{'═' * 60}\n")


# ─── CLI ──────────────────────────────────────────────────────────────────────
def main():
    parser = argparse.ArgumentParser(
        description="Campus Notifications — Priority Inbox"
    )
    parser.add_argument(
        "-n", "--top",
        type=int,
        default=10,
        help="Number of top priority notifications to display (default: 10)",
    )
    parser.add_argument(
        "--json",
        action="store_true",
        help="Output results as JSON instead of pretty-print",
    )
    args = parser.parse_args()

    print(f"[INFO] Fetching notifications from API...")
    top = get_top_n_notifications(n=args.top)

    if args.json:
        print(json.dumps(top, indent=2))
    else:
        display_priority_inbox(top, args.top)


if __name__ == "__main__":
    main()

File: notification_app_be/priorityInbox.js
/**
 * Campus Notifications - Priority Inbox Algorithm
 * Stage 1: Fetch notifications and rank top-N by weight + recency
 */

const API_URL = "http://4.224.186.213/evaluation-service/notifications";
const ACCESS_TOKEN = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJNYXBDbGFpbXMiOnsiYXVkIjoiaHR0cDovLzIwLjI0NC41Ni4xNDQvZXZhbHVhdGlvbi1zZXJ2aWNlIiwiZW1haWwiOiIyM2JxMWE0MjY2QHZ2aXQubmV0IiwiZXhwIjoxNzgwNjM4MDA2LCJpYXQiOjE3ODA2MzcxMDYsImlzcyI6IkFmZm9yZCBNZWRpY2FsIFRlY2hub2xvZ2llcyBQcml2YXRlIExpbWl0ZWQiLCJqdGkiOiIyY2JjMjUyZi1iZDU2LTRkZDMtYTc4Ny0zNDcwMmRjYmE0YjUiLCJsb2NhbGUiOiJlbi1JTiIsIm5hbWUiOiJqaWxsZWxsYSBzdXNtaXRoYSIsInN1YiI6ImEzODkzYzg2LWVlZjItNGNhYy04ZmQ2LTM1OTM5NDJmYjVhOCJ9LCJlbWFpbCI6IjIzYnExYTQyNjZAdnZpdC5uZXQiLCJuYW1lIjoiamlsbGVsbGEgc3VzbWl0aGEiLCJyb2xsTm8iOiIyM2JxMWE0MjY2IiwiYWNjZXNzQ29kZSI6IlFRZEVZeSIsImNsaWVudElEIjoiYTM4OTNjODYtZWVmMi00Y2FjLThmZDYtMzU5Mzk0MmZiNWE4IiwiY2xpZW50U2VjcmV0IjoiV0pTcmVEUkdIYUthYWVGZSJ9.ATHTGj_kYBNeqjim1CTUju0Ujfn0-TdWzlQlmC_OGpA";

// --- Weight Configuration ---
// Priority: Placement > Result > Event
const TYPE_WEIGHTS = {
  Placement: 3,
  Result: 2,
  Event: 1,
};

/**
 * Fetches all notifications from the API.
 * @returns {Promise<Array>} Array of notification objects
 */
async function fetchNotifications() {
  const response = await fetch(API_URL, {
    headers: {
      Authorization: `Bearer ${ACCESS_TOKEN}`,
    },
  });

  if (!response.ok) {
    throw new Error(`API request failed with status: ${response.status}`);
  }

  const data = await response.json();
  return data.notifications;
}

/**
 * Calculates a priority score for a single notification.
 *
 * Score formula:
 *   score = TYPE_WEIGHT * W_type + RECENCY_SCORE * W_recency
 *
 * Recency score = 1 / (1 + ageInMinutes)
 *   - A notification from NOW gets score ~1.0
 *   - A notification from 60 mins ago gets score ~0.016
 *   - This ensures fresher notifications rank higher within the same type
 *
 * Weights: type importance contributes 70%, recency contributes 30%
 *
 * @param {Object} notification
 * @returns {number} Priority score
 */
function calculatePriorityScore(notification) {
  const W_TYPE = 0.7;
  const W_RECENCY = 0.3;
  const MAX_TYPE_WEIGHT = Math.max(...Object.values(TYPE_WEIGHTS)); // normalize

  const typeWeight = (TYPE_WEIGHTS[notification.Type] || 0) / MAX_TYPE_WEIGHT;

  const timestamp = new Date(notification.Timestamp);
  const now = new Date();
  const ageInMinutes = (now - timestamp) / (1000 * 60);
  const recencyScore = 1 / (1 + ageInMinutes);

  return W_TYPE * typeWeight + W_RECENCY * recencyScore;
}

/**
 * Returns the top N priority notifications from the list.
 * @param {Array} notifications - All fetched notifications
 * @param {number} n - How many top notifications to return
 * @returns {Array} Top N notifications with score attached
 */
function getTopNPriorityNotifications(notifications, n = 10) {
  const scored = notifications.map((notification) => ({
    ...notification,
    priorityScore: calculatePriorityScore(notification),
  }));

  // Sort descending by score
  scored.sort((a, b) => b.priorityScore - a.priorityScore);

  return scored.slice(0, n);
}

/**
 * Displays the priority inbox in a readable table format.
 * @param {Array} topNotifications
 */
function displayPriorityInbox(topNotifications) {
  console.log("\n========================================");
  console.log("       CAMPUS NOTIFICATIONS - TOP 10    ");
  console.log("        Priority Inbox (Stage 1)        ");
  console.log("========================================\n");

  topNotifications.forEach((n, index) => {
    console.log(`Rank #${index + 1}`);
    console.log(`  ID        : ${n.ID}`);
    console.log(`  Type      : ${n.Type}`);
    console.log(`  Message   : ${n.Message}`);
    console.log(`  Timestamp : ${n.Timestamp}`);
    console.log(`  Score     : ${n.priorityScore.toFixed(6)}`);
    console.log("----------------------------------------");
  });

  console.log(`\nTotal notifications fetched and ranked successfully.`);
}

/**
 * Main entry point
 */
async function main() {
  try {
    const N = 10; // Number of top notifications to display (configurable: 10, 15, 20, etc.)

    console.log(`Fetching notifications from API...`);
    const notifications = await fetchNotifications();
    console.log(`Fetched ${notifications.length} notifications.`);

    console.log(`\nCalculating priority scores...`);
    const topNotifications = getTopNPriorityNotifications(notifications, N);

    displayPriorityInbox(topNotifications);
  } catch (error) {
    console.error("Error running priority inbox:", error.message);
    process.exit(1);
  }
}

main();